### PR TITLE
IGNITE-19099 Java client: Add basic authentication support

### PR DESCRIPTION
**Pull request by ptupitsyn**
* Add another client handshake extension `authn-type` to future-proof for different providers.
* Rename existing handshake extensions to be more generic and match `AuthenticationRequest` on server.
* Add `AnonymousRequest` to handle the case when client does not provide in handshake.
* Add `IgniteClientAuthenticator` interface on the client side, with a single `BasicAuthenticator` implementation. 
* Update `IgniteClientConfiguration` and pass credentials in the handshake.
* Move `AuthenticationException` from internal package to public in `ignite-api` module, add missing constructor.

### CODE REVIEW

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientInboundMessageHandler.java [hunk 1]
```diff
@@ -321,9 +324,26 @@ private void handshake(ChannelHandlerContext ctx, ClientMessageUnpacker unpacker
...
+    }
+
+    private static AuthenticationRequest<?, ?> createAuthenticationRequest(Map<HandshakeExtension, Object> extensions) {
+        Object authnType = extensions.get(HandshakeExtension.AUTHENTICATION_TYPE);
+
+        if (authnType == null) {
+            return new AnonymousRequest();
+        }
+
+        if (authnType.equals("basic")) {
```
Comment (isapego) [inline modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientInboundMessageHandler.java:340]: Probably, need to move `"basic"` to some constant.

File: modules/client/src/main/java/org/apache/ignite/client/IgniteClientAuthenticator.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.client;
+
+/**
+ * Ignite client authenticator.
+ */
```
Comment (isapego) [inline modules/client/src/main/java/org/apache/ignite/client/IgniteClientAuthenticator.java:22]: Lets add that it is used by a client to get credentials during authentication on a server node.

File: modules/client/src/main/java/org/apache/ignite/client/BasicAuthenticator.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.client;
+
+/**
+ * Basic authenticator.
+ */
```
Comment (isapego) [inline modules/client/src/main/java/org/apache/ignite/client/BasicAuthenticator.java:22]: Probably, should add what it means (single user, username + password transferred as a plain text). Probably, advice using it only for testing purposes or with TLS enabled?

**Review by isapego — APPROVED** [review]

File: modules/client/src/main/java/org/apache/ignite/client/BasicAuthenticator.java (hunk 3 above)
Comment (ptupitsyn) [reply modules/client/src/main/java/org/apache/ignite/client/BasicAuthenticator.java]: Done.

**Review by isapego — APPROVED** [review]

**Pull request merged**